static/js/compose_actions.js

```diff
@@ -98,6 +113,29 @@ function show_box(msg_type, opts) {
     set_focus(msg_type, opts);
 }
 
+let togglable_message_type;
+
+function toggle_message_type(opts) {
+    togglable_message_type = togglable_message_type === "private" ? "stream" : "private";
+    compose_state.set_message_type(togglable_message_type);
+    opts.message_type = togglable_message_type;
+    opts.trigger = "toggle message";
+    opts.stream = compose_state.stream_name();
+    opts.topic = compose_state.topic();
+    opts.private_message_recipient = compose_state.private_message_recipient();
+    reset_message_type_switcher(opts);
+    show_box(togglable_message_type, opts);
```

Comment:
It's possible the textarea height reset in `show_box` is wrong for this context. It seems possible to me that it'd be better to refactor `show_box` to have a slightly different interface (or have split out helpers e.g., "switch_to_stream_recipient").

I'm not sure what's best, but I think probably some refactoring is in order.

Reply:
I'll play around with that, thanks for calling it out as a likely refactor place!